app/drivers/zephyr/battery_voltage_divider.c

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright (c) 2020 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#define DT_DRV_COMPAT zmk_battery_voltage_divider
+
+#include <device.h>
+#include <drivers/gpio.h>
+#include <drivers/adc.h>
+#include <drivers/sensor.h>
+#include <logging/log.h>
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
+
+struct io_channel_config {
+    const char *label;
+    uint8_t channel;
+};
+
+struct gpio_channel_config {
+    const char *label;
+    uint8_t pin;
+    uint8_t flags;
+};
+
+struct bvd_config {
+    struct io_channel_config io_channel;
+    struct gpio_channel_config power_gpios;
+    uint32_t output_ohm;
+    uint32_t full_ohm;
+};
+
+struct bvd_data {
+    struct device *adc;
+    struct device *gpio;
+    struct adc_channel_cfg acc;
+    struct adc_sequence as;
+    uint16_t adc_raw;
+    uint16_t voltage;
+    uint8_t state_of_charge;
+};
+
+static uint8_t lithium_ion_mv_to_pct(int16_t bat_mv) {
+    // Simple linear approximation of a battery based off adafruit's discharge graph:
+    // https://learn.adafruit.com/li-ion-and-lipoly-batteries/voltages
+
+    if (bat_mv >= 4200) {
+        return 100;
+    } else if (bat_mv <= 3450) {
+        return 0;
+    }
+
+    return bat_mv * 2 / 15 - 459;
+}
+
+static int bvd_sample_fetch(struct device *dev, enum sensor_channel chan) {
+    struct bvd_data *drv_data = dev->driver_data;
+    const struct bvd_config *drv_cfg = dev->config_info;
+    struct adc_sequence *as = &drv_data->as;
+
+    // Make sure selected channel is supported
+    if (chan != SENSOR_CHAN_GAUGE_VOLTAGE && chan != SENSOR_CHAN_GAUGE_STATE_OF_CHARGE) {
+        return -ENOTSUP;
+    }
+
+    int rc = 0;
+
+    // Enable power GPIO if present
+    if (drv_data->gpio) {
+        rc = gpio_pin_set(drv_data->gpio, drv_cfg->power_gpios.pin, 1);
+
+        if (rc != 0) {
+            LOG_DBG("Failed to enable ADC power GPIO: %d", rc);
+            return rc;
+        }
+    }
+
+    // Read ADC
+    rc = adc_read(drv_data->adc, as);
+    as->calibrate = false;
+
+    if (rc == 0) {
+        int32_t val = drv_data->adc_raw;
+
+        adc_raw_to_millivolts(adc_ref_internal(drv_data->adc), drv_data->acc.gain, as->resolution,
+                              &val);
+
+        uint16_t millivolts = val * (uint64_t)drv_cfg->full_ohm / drv_cfg->output_ohm;
+        LOG_DBG("ADC raw %d ~ %d mV => %d mV\n", drv_data->adc_raw, val, millivolts);
+        uint8_t percent = lithium_ion_mv_to_pct(millivolts);
+        LOG_DBG("Percent: %d", percent);
+
+        drv_data->voltage = millivolts;
+        drv_data->state_of_charge = percent;
+    } else {
+        LOG_DBG("Failed to read ADC: %d", rc);
```

Comment:
While first reading this I thought that it made sense to not return an error so that the power GPIO is still disabled.
This can have a bad side effect for the application calling the `sensor_sample_fetch`.
The call to `sensor_sample_fetch` will most likely succeed due to `rc = gpio_pin_set(drv_data->gpio, drv_cfg->power_gpios.pin, 0);` below, even if `adc_read` failed.

I think it is acceptable as the `LOG_DBG` at least will indicate something is wrong, but be aware that even if the application is doing error checking, this can get undetected.

If you want to take this into account, you could introduce a second variable `rc_2` to do `rc_2 = adc_read(...);`
and in ligne 112 `return (rc_2 != 0 ? rc_2 : rc);`

Reply:
I agree we should check for both. Probably easier to just have the later code in the disable block set `int rc2 = gpio_pin_set(blah);` and check that `rc2` internally there and bail if it fails.